feat: databricks adapter

MattiasMTS opened the pull request:
# What has changed
- new adapter for databricks SQL warehouse by utilising https://github.com/databricks/databricks-sql-go
- new core structure function
- extended candies for UI

bcp-dev-ops commented:
since #80 was rejected-- what steps can I take to get this one across the finish line?

kndndrj commented:
> since #80 was rejected-- what steps can I take to get this one across the finish line?

I just don't see the point of adding a for-loop which can be done on the user side. So from my point of view, let's remove those changes from this PR.

Other than that, I'm not sure what needs to be done on the go side.

Maybe @MattiasMTS can elaborate.

MattiasMTS commented:
Ah sorry I totally missed this. 

Yea I agree with @kndndrj that the #80 commit will be removed. 

I've left some TODOs in the description. Give me some time and I can get this in. @bcp-dev-ops whats the urgency?

MattiasMTS commented:
Should be ready for review now @kndndrj.

I refactored some of the core stuff, e.g. `getPGStructure` since we are re-using this more than just the postgres adapter at this point.
Also, extended the candies a bit and added some default icons for stuff. Not sure how elegant this is given that some types only exist for some adapters. This could need some thinking but for now I don't want to create a breaking change in the lua config. Feel free to change the icons I added -> for now we separate e.g. "materialized view" from "view" by colors etc.

@bcp-dev-ops let me know if this works for you.

MattiasMTS commented on dbee/adapters/postgres_driver.go:
```diff
@@ -102,50 +101,19 @@ func (c *postgresDriver) SelectDatabase(name string) error {
+		return core.StructureTypeMaterializedView
+	case "SINK":
+		return core.StructureTypeSink
+	case "SOURCE":
```
This was added in a bit of a shadow mode to support [Risingwave](https://risingwave.com/) which is a streaming database built on top of Postgres. I hope that it alright - let me know if you want this in a separate PR then

bcp-dev-ops commented:
Works like a charm! Thanks @MattiasMTS 

@kndndrj  can you consider for merging?

phdah commented on dbee/adapters/databricks.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+// Register client
+func init() {
+	_ = register(&Databricks{}, "databricks")
+}
```
There seems to be something wrong in loading this adapter. I specify type `databricks`, but get this error:
```
[core]: failed registering source: persistence.json function DbeeCreateConnection[1]..remote#define#request, line 2: Vim(let):Error invoking '0:function:DbeeCreateConnection' on ch
annel 3:                                                                                                                                                                            
adapters.NewConnection: Mux.GetAdapters: no driver registered for provided type alias  
```
I have loaded using `lazy` as such:
```lua
    {
    	"MattiasMTS/nvim-dbee",
        event = "VeryLazy",
        -- "kndndrj/nvim-dbee",
        branch = "mattias/databricks-adapter",
        dependencies = {
            "MunifTanjim/nui.nvim",
        },
        build = function()
            require("dbee").install()
        end,
    },
```
and default `setup`.

Am i missing something here?

> Note that other adapters works as intended.

MattiasMTS replied:
Change to:

```lua
build = function()
    require("dbee").install("go")
end
```

to force installing via go. Basically because the default will do a lookup (see docs) and binaries don't exist for all branches. So, installing a specific branch require you to do it in a developer way -> which is using go.

phdah replied:
Awesome, thanks it works great now!

phdah replied:
Awesome, thanks it works great now!

phdah commented:
Noticed that there are discrepancies in when the helper functions show, in regards to `table` vs `view`. Is this a `databricks` adapter issue, or a `plugin` issue?

MattiasMTS commented on dbee/adapters/databricks.go:
```diff
@@ -0,0 +1,70 @@
+	}, nil
+}
+
+// GetHelpers returns a map of helper queries for the given table.
+func (d *Databricks) GetHelpers(opts *core.TableOptions) map[string]string {
+	// TODO: extend this to include more helper queries
+	list := fmt.Sprintf("SELECT * FROM %s.%s LIMIT 100;", opts.Schema, opts.Table)
+	columns := fmt.Sprintf(`
+SELECT *
+FROM information_schema.column
+WHERE table_schema = '%s' 
+	AND table_name = '%s';`,
+		opts.Schema, opts.Table)
+
+	return map[string]string{
+		"List":    list,
+		"Columns": columns,
+	}
+}
```
Nice! I'll try to get this in when I've time. Thanks for the comment.

phdah commented on dbee/adapters/databricks_driver.go:
```diff
@@ -0,0 +1,130 @@
+
+// getDatabricksStructureType returns the core.StructureType based on the
+// given type string for databricks adapter.
+func getDatabricksStructureType(typ string) core.StructureType {
+	switch typ {
+	case "TABLE", "BASE TABLE", "SYSTEM TABLE":
+		return core.StructureTypeTable
+	case "VIEW", "SYSTEM VIEW":
+		return core.StructureTypeView
+	case "MATERIALIZED_VIEW":
+		return core.StructureTypeMaterializedView
+	case "STREAMING_TABLE":
+		return core.StructureTypeStreamingTable
+	case "MANAGED":
+		return core.StructureTypeManaged
+	default:
+		return core.StructureTypeNone
+	}
+}
```
@MattiasMTS, these new types are not supported in the UI. The tree is, e.g., not displaying objects `"managed"`. A simple solution would be to consider there as `"table"`, which would make the core code more modular since these are in `Unity Catalog` terms considered to be just that; `tables`.
```suggestion
func getDatabricksStructureType(typ string) core.StructureType {
	switch typ {
	case "TABLE", "BASE TABLE", "SYSTEM TABLE", "MANAGED", "STREAMING_TABLE":
		return core.StructureTypeTable
	case "VIEW", "SYSTEM VIEW", "MATERIALIZED_VIEW":
		return core.StructureTypeView
	default:
		return core.StructureTypeNone
	}
}

```
This could also be handled on the core side as well, but think for standardisation reasons it makes sense to use the `StructureType`s already in place.

MattiasMTS replied:
sure I can fix this.

phdah replied:
Thanks, also noticed that there are more type, e.g, `MANAGED_SHALLOW_CLONE`, that should probably be included in the `table` category. I haven't found a complete list of table types, but an idé would be to have `*MANAGED*` be a wildcard for tables, and `*VIEW*` for view's. But The above should be sufficient for now I suppose.

MattiasMTS commented:
@kndndrj  any objections on getting this in to unblock #77 (if it is still relevant)

PyDataBlog commented:
Can we get some traction on this @kndndrj

PyDataBlog commented:
@MattiasMTS do you use this branch on your side as this project doesn't seem actively maintained?

MattiasMTS commented:
> @MattiasMTS do you use this branch on your side as this project doesn't seem actively maintained?

I haven't been working on this branch in particular but it should be sufficient to use for your daily SQL warehouse work. I'll try to fix the comments that @phdah has brought up ones this becomes available to be meged.

The project is maintained but I understand that the owner has a lot on his plate with work and other stuff going on. It's not fair to push someone to get it merged unless the person has the resources to do it. Nevertheless, it is open source.

For now, I'd say you can try and checkout this branch locally, build it and install via `go` installation and try it out.

phdah commented:
Hey! @MattiasMTS , not sure if you've heard anything from @kndndrj in terms of picking up this project again? But I was thinking if you and/or more people would like to pick up continuing maintaining this plugin if @kndndrj does not have the time.

I really like to plugin and have been using some private patched version of this branch for some time now. 

I would be willing to be an active contributor to this project, but that is bound to @kndndrj granting permission to do so.

Any thoughts and excitement? 😄

MattiasMTS commented:
>Hey! @MattiasMTS , not sure if you've heard anything from @kndndrj in terms of picking up this project again? But I was thinking if you and/or more people would like to pick up continuing maintaining this plugin if @kndndrj does not have the time.

Hi! I can absolutely think of being a contributor - perhaps not that active more than general patches being merged that I've also been using. 
Perhaps it is time to move this to a wiki discussion or something? As said earlier, I totally understand @kndndrj situation (as any other) and I don't want to push anyone into a position where they are forced to make a decision they don't want.

Ergo, let's move this discussion somewhere else -> WDYT?

phdah commented:
> > Hey! @MattiasMTS , not sure if you've heard anything from @kndndrj in terms of picking up this project again? But I was thinking if you and/or more people would like to pick up continuing maintaining this plugin if @kndndrj does not have the time.
> 
> Hi! I can absolutely think of being a contributor - perhaps not that active more than general patches being merged that I've also been using. Perhaps it is time to move this to a wiki discussion or something? As said earlier, I totally understand @kndndrj situation (as any other) and I don't want to push anyone into a position where they are forced to make a decision they don't want.
> 
> Ergo, let's move this discussion somewhere else -> WDYT?

That sound awesome! Yes, let's move this to a wiki discussion 😁

phdah commented:
@MattiasMTS , I have started a discussion here: https://github.com/kndndrj/nvim-dbee/discussions/170

MattiasMTS commented:
Thanks for initiating the discussion. I'll be back from vacation tomorrow and by then I'll try to look over getting this merged. I'll also start with the test suite for the other adapters and core tools.

Any other comments you've from your fork @phdah you want to bring up before re-review?

MattiasMTS commented:
Brushed this up a bit. Unfortunately, databricks doesn't have a docker image that this adapter could be tested on (of course), so I went with the `sqlmock` package to try out some of the simple (expected) queries against the driver.

I would appreciate any final re-reviews before merging this (e.g. @phdah) - otherwise I think we are also OK by extending this adapter in other PRs (given this is already quite big - sorry about that).

PyDataBlog commented:
> Brushed this up a bit. Unfortunately, databricks doesn't have a docker image that this adapter could be tested on (of course), so I went with the `sqlmock` package to try out some of the simple (expected) queries against the driver.
> 
> 
> 
> I would appreciate any final re-reviews before merging this (e.g. @phdah) - otherwise I think we are also OK by extending this adapter in other PRs (given this is already quite big - sorry about that).

Happy to test it locally. Been waiting for this for months

phdah commented on dbee/adapters/databricks_driver.go:
```diff
@@ -0,0 +1,119 @@
+// given type string for databricks adapter.
+func getDatabricksStructureType(typ string) core.StructureType {
+	switch typ {
+	case "TABLE", "BASE TABLE", "SYSTEM TABLE", "MANAGED", "STREAMING_TABLE", "MANAGED_SHALLOW_CLONE":
```
```suggestion
	case "TABLE", "BASE TABLE", "SYSTEM TABLE", "MANAGED", "STREAMING_TABLE", "MANAGED_SHALLOW_CLONE", "MANAGED_DEEP_CLONE":
```
Still missing this one.

phdah reviewed (approved):
I'm still missing one of the table types, but it's a hard approve otherwise 😄 Very exciting to get this merged! 🎉
